### DISCUSSION TITLE
AppleScript vs PhotoKit

### DISCUSSION THREAD

**Discussion by torarnv** [RhetTbull/osxphotos#1558] [General]
Is there a list of the remaining issues lack of feature parity between these two mechanisms of exporting photos? I would assume PhotoKit should be the more stable approach, given that it's a supported direct API, compared to controlling Photos.app via AppleScript, but perhaps that's not the case? What would it take to make PhotoKit be a stable (and preferred) export backend for `osxphotos` ? Thanks!

**Comment by RhetTbull** [comment]
I'd like to switch to PhotoKit and have a separate [project](https://github.com/RhetTbull/photokit) to create a more robust PhotoKit interface but just haven't had time to complete this and test. Several users have said they've had issues with PhotoKit. I think most of these are due to the user not granting permission when the terminal first asks for permission to access Photos. If the user does that, future calls fail.  But I've not been able to do enough testing with different configurations to figure this out.

Also, the current PhotoKit code doesn't provide a way to create albums which is a feature many users use.  I plan to add this but again, haven't had the time.

**Reply by torarnv** [reply]
Thanks @RhetTbull.

The permission issue can potentially be solved by having a small wrapper executable/script that (re)launches `osxphotos` with `responsibility_spawnattrs_setdisclaim ` as described here: https://www.qt.io/blog/the-curious-case-of-the-responsible-process

That should bring up a fresh permission dialog for `osxphotos` on first launch, instead of picking up any prior permission denials the user has done for iTerm/Terminal in the past.

**Reply by RhetTbull** [reply]
Thanks @torarnv I will take a look at that article. If I can figure out how to make osxphotos ask for permission that would definitely help!

**Reply by torarnv** [reply]
I see you are already building `osxphotos` into a standalone executable. How/where are you doing that? The easiest would be to inject something based on [this code](https://codereview.qt-project.org/c/qt-creator/qt-creator/+/386978/11/src/tools/disclaim/disclaim.mm) that would re-launch itself disclaimed. Does the tool you use to freeze the python code into an executable allow a custom `main` or injecting other sources or libraries (you could put the logic in a constructor function/static initializer)?

**Reply by torarnv** [reply]
> I see you are already building `osxphotos` into a standalone executable. How/where are you doing that? 

Never mind, found it :) Looking at a fix now

**Reply by RhetTbull** [reply]
The executable is currently built with [pyinstaller](https://pyinstaller.org) though I plan to change the build process to use [pyapp](https://github.com/ofek/pyapp) once I get done with the current round of projects (see #1369). PyInstaller effectively bundles the python interpreter and the code in a self-extracting zip that gets extracted and run at runtime. PyApp uses rust to build a custom executable that when first run downloads Python and installs the source code then replaces itself with the python process. I've started using PyApp for my other projects and it's a better user experience as it's a smaller executable (because Python is not downloaded until the first time it's run) and can upgrade the osxphotos version after it's installed.  


Whatever fix would almost certainly need to be written in Python to work with either of these installers.

**Reply by torarnv** [reply]
Ah, good to know. The approach is fairly generic, so hopefully it integrates easily with `pyapp` as well. A quick prototype with a custom runtime hook for `pyinstaller` gives me:

<img width="1035" alt="image" src="https://github.com/RhetTbull/osxphotos/assets/7536/8581764b-682d-4eec-b559-b8191e847ec3">

**Reply by torarnv** [reply]
<img width="501" alt="image" src="https://github.com/RhetTbull/osxphotos/assets/7536/58e19d82-84cb-49f1-b413-414907037a39">

**Reply by RhetTbull** [reply]
Excellent! This should also help with terminals that don't include the `NSPhotoLibraryUsageDescription` in their Info.plist. For example, some versions of iTerm2 include a regression that drops this preventing osxphotos from working as the terminal never asks for permission.

**Reply by RhetTbull** [reply]
Hi @torarnv do you have example code for your prototype pyinstaller hook you can share? I'd like to explore getting something similar working with pyapp for another project.

**Reply by torarnv** [reply]
Sure! I'll push a PR today or tomorrow if that's okey?

**Reply by torarnv** [reply]
Hold tight! 😃

**Reply by torarnv** [reply]
Here you go: https://github.com/RhetTbull/osxphotos/pull/1587

**Reply by RhetTbull** [reply]
@torarnv Thank you! This is great work and it will be much better for the user to have osxphotos manage the permissions than relying on the Terminal. One thing I've got figure out is asking at the right time so osxphotos doesn't fail. For example, when running:

`./osxphotos export ~/Desktop/export --verbose --use-photos-export --use-photokit`

The first time, the dialog box pops up asking permission to the Photos library. But in the background, osxphotos continues to run and fails with a permission error because user hasn't click "grant access" yet. 

```
osxphotos version: 0.68.2
Python version: 3.12.1 (v3.12.1:2305ca5144, Dec  7 2023, 17:23:39) [Clang 13.0.0 (clang-1300.0.29.30)]
Platform: macOS-13.5.1-x86_64-i386-64bit, 13.5.1
Verbose level: 1
Exporting 21 photos to /Users/rhet/Desktop/export...
Exporting Pumkins1.jpg (F12384F6-CD17-4151-ACBA-AE0E3688539E.jpeg) (1/21)
Error exporting photo (F12384F6-CD17-4151-ACBA-AE0E3688539E: Pumkins1.jpg) as Pumkins1.jpg: Could not get authorizaton to use Photos: auth_status = 0
```

This might work:
If the auth check fails, tell user they need to grant permission, start a loop that sleeps, checks again, and do this X times. If it fails after X times, abort with a message that osxphotos could not get permissions and guide user how to do this in Settings > Privacy & Security.

**Reply by RhetTbull** [reply]
One additional issue: I can't figure out how to reset the permissions with tccutil as the command line executable lacks a bundle ID. Some info on [this thread](https://forums.developer.apple.com/forums/thread/697278). Short of blowing away the entire TCC database or running [tccutil.py](https://github.com/jacobsalmela/tccutil), which requires SIP disabled, I don't see a way to reset permissions which is needed for testing. Additionally, it appears tccutil uses path if there's no bundle ID so that means running two copies of osxphotos from two different paths would result in duplicate entries in the TCC database which could also call confusion. 

Will need to do some more testing before rolling this out.

@torarnv are you aware of any way to reset the TCC permissions when using the disclaim method?

**Reply by torarnv** [reply]
> @torarnv are you aware of any way to reset the TCC permissions when using the disclaim method?

For testing you can use `tccutil reset Photos`, without a bundle ID. It will affect all apps, but for testing that might be acceptable.

> Additionally, it appears tccutil uses path if there's no bundle ID so that means running two copies of osxphotos from two different paths would result in duplicate entries in the TCC database which could also call confusion.

It's possible to associate a bundle ID with a standalone executable by embedding an Info.plist in the executable itself, if you pass `-sectcreate __TEXT __info_plist Info.plist` to the linker. For `pyinstaller` this would require building a custom version of the bootloader AFAICT. Alternatively you can build `osxphotos` into an app bundle via `pyinstaller`, in which case you can pass a custom Info.plist with a bundle ID: https://pyinstaller.org/en/stable/spec-files.html#spec-file-options-for-a-macos-bundle

Before shipping this I would also verify that launching osxphotos (as distributed to users, i.e. signed) repeatedly will only ask for permission the first time. Sometimes the TCC machinery gets confused and doesn't persist the user's action if the signature or bundle ID is missing, so good to verify that.

**Reply by torarnv** [reply]
> It's possible to associate a bundle ID with a standalone executable by embedding an Info.plist in the executable itself, if you pass `-sectcreate __TEXT __info_plist Info.plist` to the linker. For `pyinstaller` this would require building a custom version of the bootloader AFAICT. 

Alternatively, you can do something like this to hot-patch the bootloader after `pyinstaller` has done its thing, but before signing: https://stackoverflow.com/a/76850148

**Reply by torarnv** [reply]
> Alternatively, you can do something like this to hot-patch the bootloader after `pyinstaller` has done its thing, but before signing: https://stackoverflow.com/a/76850148

Or `brew install llvm` and use https://llvm.org/docs/CommandGuide/llvm-objcopy.html#cmdoption-llvm-objcopy-add-section

**Reply by RhetTbull** [reply]
Thanks @torarnv I'll read up on these ideas. My ultimate goal is to figure out how to make your solution work with [pyapp](https://github.com/ofek/pyapp) which would enable osxphotos to self update when there's a new version. PyApp creates an executable built in Rust that at first execution, downloads python and the osxphotos package then spawns a new process to run osxphotos. In subsequent runs, the previously installed python executable is spawned to run osxphotos. There's likely a way to make this work but it'll be several weeks before I have time to dive into PyApp internals.

**Reply by RhetTbull** [reply]
I think I've got this working for both pyinstaller and pyapp. Will need to do a new release before I can fully test it (pyapp pulls from PyPI to build the exe). I moved the `disclaim()` call to the main cli entry point and check at that time for pyinstaller / pyapp. This should allow osxphotos to work as it currently does (request permission through terminal) when called from python but to use the disclaim call when called from an executable.

**Reply by torarnv** [reply]
Awesome @RhetTbull ! Do you have the patch for pyapp that I can take a look at?

**Reply by RhetTbull** [reply]
I've got it on a separate branch at the moment. See new [disclaim.py](https://github.com/RhetTbull/osxphotos/blob/feature_disclaim/osxphotos/disclaim.py) and where it's called in the cli here:

 https://github.com/RhetTbull/osxphotos/blob/9333cf306802c14bdedb14cd044e460ec48d1c29/osxphotos/cli/cli.py#L111-L113

**Reply by torarnv** [reply]
Thanks! Look forward to testing it out!

**Reply by RhetTbull** [reply]
I've gotten pyapp working to reliably build the executable and using my [applecrate](https://github.com/RhetTbull/applecrate) tool to package it into a signed macOS installer which I should eventually be able to also handle notarization. This would vastly simplify distribution of osxphotos. Additionally, it has built in update so you can run `osxphotos self update` to install the latest version. This all works great with one big caveat: due to the way pyapp works the process that actually gets spawned is `python3.13` not `osxphotos` so the dialog box shows `python3.13` requesting permission and that's also what shows up in the Privacy & Security settings.

<img width="269" alt="Screenshot 2024-12-04 at 9 34 19 PM" src="https://github.com/user-attachments/assets/ae37a454-fc0f-4561-ac42-e0eaefa6517c">

<img width="712" alt="Screenshot 2024-12-04 at 9 34 49 PM" src="https://github.com/user-attachments/assets/57767599-664d-4e43-a075-42daa4017882">

The pyapp built app runs much faster than the pyinstaller version (10x on initial load) and has the nice self-update feature but I don't know if this trade-off is worth it as seeing `python3.13` (or whatever the python executable is) in the permission dialog may be confusing. Will need to tinker with this and think about it some more.

I have an idea for how to make the pyinstaller version also self-updating but need to play with that. There's still the issue that the pyinstaller version (which is a self-extracting zip) is very slow...on my 2019 Intel Macbook Air, it takes ~40 seconds to load every time `osxphotos` is run.

**Reply by rajscode** [reply]
Could you wrapper everything so the final installed item when it executes calls an Automator script that is a simple launcher (i.e. osxphotos_automator_script.app)? That way you permission the osxphotos_automator_script.app one time with privelege and it doesn't need to itself ever be updated whilst the underlying items can be updated without MacOS having priv pop ups happening again and again. i.e. https://github.com/RhetTbull/osxphotos/discussions/1080#discussioncomment-8561339

**Reply by RhetTbull** [reply]
@rajscode that's a possibility but it would need to know where the python executable and script files are which somewhat complicates installation. I haven't tried this but I think it would also mean that launching the app would require "open osxphotos" as opposed to just typing osxphotos on the command line.  I may default back to having the Terminal handle permissions but putting in code in osxphotos to verify it has the permissions and prompting the user what to do if Terminal has granted them.
